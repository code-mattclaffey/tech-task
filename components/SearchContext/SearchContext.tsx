import React, { useState, useContext } from "react";
import { getLocations, getLatestParametersFromCity } from "../../apis/api";
import { LocationsContext } from '../LocationsContext/LocationsContext';


export const SearchContext = React.createContext({
    searchTerm: "",
    searchResults: [],
    setSearchTerm: (searchTerm: string) => searchTerm,
    handleKeyEvent: (event: any) => event,
    resetSearchResults: () => null,
    selectLocation: (city: string) => city
});

type ProviderProps = {
    searchResults?: Array<any>;
};

export const SearchContextProvider: React.FunctionComponent<ProviderProps> = ({
    children,
    searchResults,
}) => {
    const { setSelectedLocations } = useContext(LocationsContext);

    const [state, setState] = useState({
        searchTerm: "",
        searchResults: searchResults || []
    });

    const reduceDulplicateCities = (allCities: any, currentCity: any) => {
        const { city } = currentCity;

        const cityAlreadyExistsInObject = allCities.filter((cityObj: any) => {
            return cityObj.city === city;
        });

        if (cityAlreadyExistsInObject.length === 0) {
            allCities.push(currentCity);
        }

        return allCities;
    };

    const setSearchTerm = (searchTerm: string) => {
        setState({ ...state, searchTerm });
    };

    const handleKeyEvent = () => {
        const { searchTerm } = state;
        if (searchTerm.length === 0) {
            setState({ ...state, searchResults: [] });
            return;
        }

        getLocations(searchTerm).then(({ results }) => {
            const searchResults = results.reduce(reduceDulplicateCities, []);
            setState({ ...state, searchResults });
        });
    };

    const resetSearchResults = () => {
        setState({ ...state, searchResults: [] });
    };

    const selectLocation = (city: string) => {
        getLatestParametersFromCity(city).then(({ results }) => {
            setSelectedLocations(results);
            setState({ ...state, searchResults: [] });
        });
    };

    const data = {
        ...state,
        setSearchTerm,
        handleKeyEvent: handleKeyEvent,
        resetSearchResults,
        selectLocation
    };

    return (
        // @ts-ignore
        <SearchContext.Provider value={data}>
            {children}
        </SearchContext.Provider>
    );
};
