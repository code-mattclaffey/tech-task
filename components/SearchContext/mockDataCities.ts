

export const mockData = {"meta":{"name":"openaq-api","license":"CC BY 4.0","website":"https://docs.openaq.org/","page":1,"limit":100,"found":7},"results":[{"id":"GB-28","country":"GB","city":"Manchester","cities":["Manchester"],"location":"Manchester Piccadilly","locations":["Manchester Piccadilly"],"sourceName":"DEFRA","sourceNames":["DEFRA"],"sourceType":"government","sourceTypes":["government"],"coordinates":{"longitude":-2.23788,"latitude":53.48152},"firstUpdated":"2016-02-27T21:00:00.000Z","lastUpdated":"2020-01-18T13:00:00.000Z","parameters":["no2","o3","pm10","pm25","so2"],"countsByMeasurement":[{"parameter":"no2","count":26469},{"parameter":"o3","count":28442},{"parameter":"pm10","count":265},{"parameter":"pm25","count":27504},{"parameter":"so2","count":27380}],"count":110060},{"id":"GB-152","country":"GB","city":"Manchester","cities":["Manchester"],"location":"Manchester Sharston","locations":["Manchester Sharston"],"sourceName":"DEFRA","sourceNames":["DEFRA"],"sourceType":"government","sourceTypes":["government"],"coordinates":{"longitude":-2.23922,"latitude":53.37131},"firstUpdated":"2016-05-04T15:00:00.000Z","lastUpdated":"2020-01-18T13:00:00.000Z","parameters":["no2","o3"],"countsByMeasurement":[{"parameter":"no2","count":27470},{"parameter":"o3","count":27447}],"count":54917},{"id":"GB-137","country":"GB","city":"Manchester","cities":["Manchester"],"location":"Salford Eccles","locations":["Salford Eccles"],"sourceName":"DEFRA","sourceNames":["DEFRA"],"sourceType":"government","sourceTypes":["government"],"coordinates":{"longitude":-2.33414,"latitude":53.48481},"firstUpdated":"2016-02-27T21:00:00.000Z","lastUpdated":"2020-01-18T13:00:00.000Z","parameters":["no2","pm10","pm25"],"countsByMeasurement":[{"parameter":"no2","count":27921},{"parameter":"pm10","count":27574},{"parameter":"pm25","count":27696}],"count":83191},{"id":"US-1977","country":"US","city":"Manchester-Nashua","cities":["Manchester-Nashua"],"location":"Miller State Park","locations":["Miller State Park"],"sourceName":"AirNow","sourceNames":["AirNow"],"sourceType":"government","sourceTypes":["government"],"coordinates":{"longitude":-71.87862,"latitude":42.86183},"firstUpdated":"2017-01-18T17:00:00.000Z","lastUpdated":"2020-01-18T12:00:00.000Z","parameters":["co","o3","pm10","pm25","so2"],"countsByMeasurement":[{"parameter":"co","count":19253},{"parameter":"o3","count":18861},{"parameter":"pm10","count":19489},{"parameter":"pm25","count":19573},{"parameter":"so2","count":18403}],"count":95579},{"id":"US-815","country":"US","city":"Manchester-Nashua","cities":["Manchester-Nashua"],"location":"Miller State Park","locations":["Miller State Park"],"sourceName":"AirNow","sourceNames":["AirNow"],"sourceType":"government","sourceTypes":["government"],"coordinates":{"longitude":-71.8789,"latitude":42.8619},"firstUpdated":"2016-03-06T19:00:00.000Z","lastUpdated":"2017-01-18T16:00:00.000Z","parameters":["co","o3","pm10","pm25","so2"],"countsByMeasurement":[{"parameter":"co","count":7141},{"parameter":"o3","count":7170},{"parameter":"pm10","count":3114},{"parameter":"pm25","count":7168},{"parameter":"so2","count":7063}],"count":31656},{"id":"US-1673","country":"US","city":"Manchester-Nashua","cities":["Manchester-Nashua"],"location":"Nashua - Gilson Rd.","locations":["Nashua - Gilson Rd."],"sourceName":"AirNow","sourceNames":["AirNow"],"sourceType":"government","sourceTypes":["government"],"coordinates":{"longitude":-71.445,"latitude":42.7019},"firstUpdated":"2016-04-13T12:00:00.000Z","lastUpdated":"2016-10-13T12:00:00.000Z","parameters":["o3"],"countsByMeasurement":[{"parameter":"o3","count":4182}],"count":4182},{"id":"US-1990","country":"US","city":"Manchester-Nashua","cities":["Manchester-Nashua"],"location":"Nashua - Gilson Rd.","locations":["Nashua - Gilson Rd."],"sourceName":"AirNow","sourceNames":["AirNow"],"sourceType":"government","sourceTypes":["government"],"coordinates":{"longitude":-71.52242,"latitude":42.71865},"firstUpdated":"2017-02-22T19:00:00.000Z","lastUpdated":"2019-10-09T08:00:00.000Z","parameters":["o3"],"countsByMeasurement":[{"parameter":"o3","count":10536}],"count":10536}]};