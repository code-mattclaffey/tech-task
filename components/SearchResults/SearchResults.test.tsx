import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import { SearchResults } from ".";
import { SearchContextProvider } from "../SearchContext";

afterEach(() => {
    cleanup();
});

describe("<SearchSuggestions />", () => {
    it("should match snapshot", () => {
        const { container } = render(
            <SearchContextProvider
                searchResults={[
                    {
                        city: "Amsterdam",
                        country: "NL",
                        count: 21301,
                        locations: 14
                    },
                    {
                        city: "Badhoevedorp",
                        country: "NL",
                        count: 2326,
                        locations: 1
                    }
                ]}
            >
                <SearchResults />
            </SearchContextProvider>
        );

        expect(container).toMatchSnapshot();
    });

    it("should read out the total amount of results", () => {
        const { queryByTestId } = render(
            <SearchContextProvider
                searchResults={[
                    {
                        city: "Amsterdam",
                        country: "NL",
                        count: 21301,
                        locations: 14
                    },
                    {
                        city: "Badhoevedorp",
                        country: "NL",
                        count: 2326,
                        locations: 1
                    }
                ]}
            >
                <SearchResults />
            </SearchContextProvider>
        );

        expect(queryByTestId('search-results-aria-live').textContent).toEqual('There are 2 suggestions. Use the up and down arrows to browse.');
        expect(queryByTestId('search-results-aria-live').hasAttribute('aria-live')).toEqual(true);
        expect(queryByTestId('search-results-aria-live').hasAttribute('role')).toEqual(true);
    });

    it("should allow the user the navigate through the options with the arrow keys", () => {
        const { queryByTestId } = render(
            <SearchContextProvider
                searchResults={[
                    {
                        city: "Amsterdam",
                        country: "NL",
                        count: 21301,
                        locations: 14
                    },
                    {
                        city: "Badhoevedorp",
                        country: "NL",
                        count: 2326,
                        locations: 1
                    }
                ]}
            >
                <SearchResults />
            </SearchContextProvider>
        );

        const searchResults = queryByTestId('search-results');

        searchResults.focus();

        fireEvent.keyDown(searchResults, { key: 'ArrowDown', keyCode: 40 });

        expect(queryByTestId('search-option-1').classList.contains('c-search-suggestions__option--focus')).toEqual(true);
    });

    it("should set the focus class on the first option when the ArrowDown key is fired on the last option", () => {
        const { queryByTestId } = render(
            <SearchContextProvider
                searchResults={[
                    {
                        city: "Amsterdam",
                        country: "NL",
                        count: 21301,
                        locations: 14
                    },
                    {
                        city: "Badhoevedorp",
                        country: "NL",
                        count: 2326,
                        locations: 1
                    }
                ]}
            >
                <SearchResults />
            </SearchContextProvider>
        );

        const searchResults = queryByTestId('search-results');

        searchResults.focus();

        fireEvent.keyDown(searchResults, { key: 'ArrowDown', keyCode: 40 });
        fireEvent.keyDown(searchResults, { key: 'ArrowDown', keyCode: 40 });

        expect(queryByTestId('search-option-0').classList.contains('c-search-suggestions__option--focus')).toEqual(true);
    });

    it("should set the focus class on the last option when the ArrowUp key is fired on the first option", () => {
        const { queryByTestId } = render(
            <SearchContextProvider
                searchResults={[
                    {
                        city: "Amsterdam",
                        country: "NL",
                        count: 21301,
                        locations: 14
                    },
                    {
                        city: "Badhoevedorp",
                        country: "NL",
                        count: 2326,
                        locations: 1
                    }
                ]}
            >
                <SearchResults />
            </SearchContextProvider>
        );

        const searchResults = queryByTestId('search-results');

        searchResults.focus();

        fireEvent.keyDown(searchResults, { key: 'ArrowUp', keyCode: 38 });

        expect(queryByTestId('search-option-1').classList.contains('c-search-suggestions__option--focus')).toEqual(true);
    });

    it("should reset the search results when escape key is selected", () => {
        const { queryByTestId } = render(
            <SearchContextProvider
                searchResults={[
                    {
                        city: "Amsterdam",
                        country: "NL",
                        count: 21301,
                        locations: 14
                    },
                    {
                        city: "Badhoevedorp",
                        country: "NL",
                        count: 2326,
                        locations: 1
                    }
                ]}
            >
                <SearchResults />
            </SearchContextProvider>
        );

        const searchResults = queryByTestId('search-results');

        searchResults.focus();

        fireEvent.keyDown(searchResults, { key: 'Escape', keyCode: 27 });

        expect(queryByTestId('search-option01')).toBeNull();
        expect(queryByTestId('search-option-1')).toBeNull();
    });
});
