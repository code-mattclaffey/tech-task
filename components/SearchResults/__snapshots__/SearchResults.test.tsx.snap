// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchSuggestions /> should match snapshot 1`] = `
<div>
  <div
    aria-live="polite"
    class="u-visually-hidden"
    data-testid="search-results-aria-live"
    role="status"
  >
    There are 2 suggestions. Use the up and down arrows to browse.
  </div>
  <div
    aria-activedescendant="Amsterdam-0"
    class="c-search-suggestions__results"
    data-testid="search-results"
    id="search-results"
    role="listbox"
    tabindex="0"
  >
    <div
      class="c-search-suggestions__option c-search-suggestions__option--focus"
      data-testid="search-option-0"
      id="Amsterdam-0"
      role="option"
      tabindex="-1"
    >
      Amsterdam
    </div>
    <div
      class="c-search-suggestions__option"
      data-testid="search-option-1"
      id="Badhoevedorp-1"
      role="option"
      tabindex="-1"
    >
      Badhoevedorp
    </div>
  </div>
</div>
`;
